<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="Content-Style-Type" content="text/css" />
  <meta name="generator" content="pandoc" />
  <title></title>
  <style type="text/css">code{white-space: pre;}</style>
  <style type="text/css">
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; }
code > span.dt { color: #902000; }
code > span.dv { color: #40a070; }
code > span.bn { color: #40a070; }
code > span.fl { color: #40a070; }
code > span.ch { color: #4070a0; }
code > span.st { color: #4070a0; }
code > span.co { color: #60a0b0; font-style: italic; }
code > span.ot { color: #007020; }
code > span.al { color: #ff0000; font-weight: bold; }
code > span.fu { color: #06287e; }
code > span.er { color: #ff0000; font-weight: bold; }
  </style>
</head>
<body>
<h1 id="bio721p-genome-bioinformatics-practical-read-mapping-and-variant-calling">BIO721P Genome-Bioinformatics practical: read mapping and variant calling</h1>
<p>Roddy Pracana and Yannick Wurm</p>
<h2 id="introduction">Introduction</h2>
<p>In the previous practical, we created a genome assembly from a single individual. But studying the genome of a single individual is rarely interesting. In most medical and evolutionary genetics studies, we actually want to study the genomes of many individuals. By studying the differences between the genomes, we can start understanding the genetic basis of phenotypic variation (such as the genetic basis of complex traits and diseases) or we can infer the evolutionary processes that have affected the population we're studying (such as understanding population structure, speciation, selection, etc.).</p>
<p>Thus, we need a method of finding genetic variants between individuals in a sensitive and accurate way, a process normally known as variant calling.</p>
<p>There are several types of variants. Commonly, people look at single nucleotide polymorphisms (SNPs, sometimes also known as single nucleotide variants, SNVs). Other classes include small insertions and deletions (known collectively as indels), as well as larger structural variants, such as large insertions, deletions, inversions and translocations.</p>
<p>There are several approaches to variant calling from short pair-end reads. We are going to use one of them. First, we are going to map the reads from each individual to a reference assembly similar to the one created in the previous practical. Then we are going to find the positions where at least some of the individuals differ from the reference (and each other).</p>
<h2 id="the-data">The data</h2>
<p>We will be analysing subsets of whole-genome sequences of several fire ant individuals. The fire ant, <em>Solenopsis invicta</em>, is notable for being dimorphic in terms of colony organisation, with some colonies having one queen and other colonies having multiple queens. Interestingly, this trait is genetically determined. In this practical, we are going to try to find the genetic difference between ants from single queen and multiple queen colonies.</p>
<p>We will be using a subset of the reads from whole-genome sequencing of 14 male fire ants. Samples 1B to 7B are from single-queen colonies, samples 1b to 7b are from multiple-queen colonies. Ants are haplodiploid, which means that they have haploid males.</p>
<p>We will align the reads to a subset of the reference genome assembly of the species (the same regions we tried to assemble earlier) using the aligner <code>bowtie2</code>. We will try to find positions that differ between each individual and the reference with the software <code>samtools</code> and <code>bcftools</code>.</p>
<h2 id="before-we-start-the-analysis">Before we start the analysis</h2>
<h3 id="connect-to-the-cluster-and-load-the-modules-we-will-be-using">Connect to the cluster and load the modules we will be using</h3>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># First, login to Apocrita</span>
<span class="kw">ssh</span> abc123@login.hpc.qmul.ac.uk

<span class="co"># From there, log into the MSc machine:</span>
<span class="kw">ssh</span> dn36 <span class="co"># or ssh dn42</span></code></pre>
<p>Now type into the console:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">module</span> load samtools/1.1 bcftools/1.1 HTSlib/1.1</code></pre>
<p>This loads the right version of the tools we are using in this practical. You need to type this if you restart a new session on the cluster, otherwise you will be using the wrong version of the tools and the practical will not work!</p>
<p>To find if you've loaded the correct version of the modules, type:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">which</span> samtools</code></pre>
<p>If you don't get version <code>1.1</code>, then you don't have the right module loaded.</p>
<h3 id="create-a-directory-for-your-analysis">Create a directory for your analysis</h3>
<p>To create a directory for the analysis in this practical, type, in your home directory:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">mkdir</span> 2015-10-map-call-practical</code></pre>
<p>Use <code>cd</code> to go into that directory.</p>
<p>There, use <code>mkdir</code> to create a directory called <code>data</code> and one called <code>2015-10-07-analysis</code>.</p>
<h3 id="create-soft-links-for-the-data-in-the-new-directory">Create soft links for the data in the new directory</h3>
<p>The input data is stored in a folder called:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">/data/SBCS-MSc-BioInf/2015-10-practical/map_call</span></code></pre>
<p>In it, you can find a file with the reference assembly (<code>reference.fa</code>) and a directory with the reads in the <code>.fq</code> format (<code>reads</code>). Make a soft link to those files from the new <code>data/</code> directory:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cd</span> ~/2015-10-map-call-practical/data
<span class="kw">ln</span> -sf /data/SBCS-MSc-BioInf/2015-10-practical/map_call/reference.fa .
<span class="kw">ln</span> -sf /data/SBCS-MSc-BioInf/2015-10-practical/map_call/reads/* .</code></pre>
<p>By doing this, you create a soft link to the data instead of copying it, so you don't multiply the amount of space you are using up. Never use the command <code>rm</code> on soft links, as it will also delete the original data. Instead, use the command <code>unlink</code>.</p>
<p>You should see the reads for 14 samples.</p>
<h4 id="q.-how-many-scaffolds-are-there-in-the-assembly-try-using-grep">Q. How many scaffolds are there in the assembly? (Try using <code>grep</code>)</h4>
<h4 id="q.-important-why-does-each-sample-have-two-sets-of-reads">Q. (important!) Why does each sample have two sets of reads?</h4>
<h4 id="q.-what-is-each-line-of-the-.fq-file">Q. What is each line of the <code>.fq</code> file?</h4>
<h4 id="q.-how-many-reads-do-we-have-in-individual-f1_b">Q. How many reads do we have in individual f1_B?</h4>
<h4 id="q.-whats-the-size-of-each-read-all-reads-have-equal-size">Q. What's the size of each read (all reads have equal size)?</h4>
<h4 id="q.-knowing-that-each-scaffold-is-200kb-what-is-the-expected-coverage-per-base-pair-of-individual-f1_b">Q. Knowing that each scaffold is 200kb, what is the expected coverage per base pair of individual f1_B?</h4>
<h3 id="create-soft-links-from-the-data">Create soft links from the data</h3>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Go back to the original directory:</span>
<span class="kw">cd</span> ~/2015-10-map-call-practical/2015-10-07-analysis
<span class="kw">ln</span> -sf ../data/reference.fa .
<span class="kw">ln</span> -sf ../data/*.fq .</code></pre>
<p>Note that the star <code>*</code> refers to all the files in the <code>../data/reads/</code> directory. The <code>.</code> is the current directory.</p>
<h2 id="aligning-reads-to-a-reference-assembly">Aligning reads to a reference assembly</h2>
<p>The first step in our pipeline is to align the paired end reads to the reference genome. We are using the software <code>bowtie2</code>, which was created to align short read sequences to long sequences such as the scaffolds in a reference assembly. <code>bowtie2</code>, like most aligners, works in two steps. In the first step, the scaffold sequence (sometimes known as the database) is indexed, in this case using the Burrows-Wheeler Transform, which allows for memory efficient alignment. The second step is the alignment itself.</p>
<p>Let's start by creating the reference index (make sure you are using <code>bowtie2</code> and not the earlier version, bowtie):</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bowtie2-build</span> reference.fa reference_index</code></pre>
<p>Now the alignment step:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bowtie2</span> -x reference_index -1 f1_B.1.fq -2 f1_B.2.fq <span class="kw">&gt;</span> f1_B.sam</code></pre>
<h4 id="q.-what-do-the-parameter--x--1-and--2-mean">Q. What do the parameter <code>-x</code>, <code>-1</code> and <code>-2</code> mean?</h4>
<p>The command produced a SAM file (Sequence Alignment/Map file), which is the standard file used to store sequence alignments. Have a quick look at the file by typing <code>less f1.sam</code>. The file includes a header (lines starting with the <code>@</code> symbol), and a line for every read aligned to the reference assembly. For each read, we are given a mapping quality values, the position of both pairs, the actual sequence and its quality by base pair, and a series of flags with additional measures of mapping quality.</p>
<p>We now need to run bowtie2 for all the other samples. We could do this by typing the same command another 13 times (changing the sample name), or we can use the <code>GNU parallel</code> tool:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## Create a file with all sample names</span>
<span class="kw">ls</span> *fq <span class="kw">|</span> <span class="kw">cut</span> -d <span class="st">&#39;.&#39;</span> -f 1 <span class="kw">|</span> <span class="kw">sort</span> <span class="kw">|</span> <span class="kw">uniq</span> <span class="kw">&gt;</span> names.txt

<span class="co">## Run bowtie with each sample (will take a few minutes)</span>
<span class="kw">parallel</span> <span class="st">&quot;bowtie2 -x reference_index -1 {}.1.fq -2 {}.2.fq &gt; {}.sam&quot;</span> :::: names.txt</code></pre>
<p>Because SAM files include a lot of information, they tend to occupy a lot of space (even in our case). Therefore, SAM files are generally compressed into BAM files (Binary sAM). Most tools that use aligned reads requires BAM files that have been sorted and indexed by genomic position. This is done using <code>samtools</code>, a set tools create to manipulate SAM/BAM files. To compress and sort a SAM file for a given sample, we type:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co">## SAM to BAM.</span>
<span class="co"># samtools view: compresses the SAM to BAM</span>
<span class="co"># samtools sort: sorts by scaffold position (creates f1_B.sorted.bam)</span>
<span class="kw">samtools</span> view -Sb f1_B.sam <span class="kw">|</span> <span class="kw">samtools</span> sort - f1_B.sorted
<span class="co">## This creates a file (f1_B.sorted.bam), which we then index</span>
<span class="kw">samtools</span> index f1_B.sorted.bam   <span class="co"># creates f1_B.sorted.bam.bai</span></code></pre>
<h4 id="q.-what-does-the-dash---in-the-samtools-sort-command-mean">Q. What does the dash (<code>-</code>) in the <code>samtools sort</code> command mean?</h4>
<p>Again, we can use parallel to run this step for all the samples:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">parallel</span> <span class="st">&quot;samtools view -Sb {}.sam | samtools sort - {}.sorted&quot;</span> :::: names.txt
<span class="kw">parallel</span> <span class="st">&quot;samtools index {}.sorted.bam&quot;</span> :::: names.txt</code></pre>
<h2 id="variant-calling">Variant calling</h2>
<p>There are several approaches to call variants. The simplest approach is to look for positions where the mapped reads consistently have a different base than the reference assembly (the consensus approach). We need to run two steps, <code>samtools mpileup</code>, which looks for inconsistencies between the reference and the aligned reads, and <code>bcftools call</code>, which interprets them as variants.</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># Step 1: samtools mpileup</span>
<span class="co">## Create index of the reference (different from that used by bowtie2)</span>
<span class="kw">samtools</span> faidx reference.fa

<span class="co"># Run samtools mpileup</span>
<span class="kw">samtools</span> mpileup -uf reference.fa *sorted.bam <span class="kw">&gt;</span> raw_calls.bcf</code></pre>
<h4 id="q.-what-does-the-symbol-mean-here">Q. What does the symbol<code>*</code> mean here?</h4>
<p>If you type <code>bcftools</code> in the console, you will see that this programme has a series of tools to manipulate VCF/BCF files. The tool we want to use is <code>bcftools call</code>, which does SNP and indel calling. As well as the consensus caller (option <code>-c</code>), which we are using, bcftools includes the multiallelic caller (option <code>-m</code>). Because we have a relatively small number of samples and low coverage for each of the sample, the consensus caller will do. Before we carry on, we need to remember that we are analysing male ants, which have haploid genomes! Because of this, we need to use option -S, which allows the user to add a list with two columns, the first with the name of the samples (i.e. that name of the <code>bam</code> files), the second with their ploidy (1 or 2).</p>
<h4 id="q.-make-a-file-called-ploidy_per_sample.txt-with-a-column-with-the-bam-file-names-e.g.-f1_b.sorted.bam-and-a-column-with-the-ploidy-level-i.e.-1.">Q. Make a file (called ploidy_per_sample.txt) with a column with the bam file names (e.g. f1_B.sorted.bam) and a column with the ploidy level (i.e. 1).</h4>
<p>Now we run the bcftools call command and save the result into a VCF file.</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bcftools</span> call -c raw_calls.bcf -v -S ploidy_per_sample.txt <span class="kw">&gt;</span> variant_calls.vcf</code></pre>
<h4 id="q.-what-does-the-parameter--v-do-under-what-situation-would-you-leave-it-out">Q. What does the parameter <code>-v</code> do? Under what situation would you leave it out?</h4>
<p>The file produced a VCF (Variant Call Format) format telling the position, nature and quality of the called variants. The VCF format is the standard format used</p>
<p>Let's take a look at the VCF file produced by typing <code>less -S variant_calls.vcf</code>. The file is composed of a header and of and rows for all the variant positions. Have a look at the different columns and check what each is (the header includes labels). Notice that some columns include several fields.</p>
<h4 id="q.-use-less-to-look-at-the-vcf-file.-where-does-the-header-start-and-end">Q. Use <code>less</code> to look at the VCF file. Where does the Header start and end?</h4>
<h4 id="q.-how-is-the-genotype-of-each-sample-coded">Q. How is the genotype of each sample coded?</h4>
<h4 id="q.-how-many-variants-were-identified">Q. How many variants were identified?</h4>
<h4 id="q.-can-you-tell-the-difference-between-snps-and-indels-how-many-of-each-have-been-identified">Q. Can you tell the difference between SNPs and indels? How many of each have been identified?</h4>
<h2 id="quality-filtering-of-variant-calls">Quality filtering of variant calls</h2>
<p>Not all variants that we called are necessarily of good quality, so it is essential to have a quality filter step. The VCF includes several fields with quality information. The most obvious is the column QUAL, which gives us a Phred-scale quality score.</p>
<h4 id="q.-what-does-a-phred-scale-quality-score-of-30-mean">Q. What does a Phred-scale quality score of 30 mean?</h4>
<p>We will filter the VCF using <code>bcftools filter</code>. We can remove anything with quality call smaller than 30:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">bcftools</span> filter --exclude <span class="st">&#39;QUAL &lt; 30&#39;</span> variant_calls.vcf <span class="kw">&gt;</span> filtered_variant_calls.vcf</code></pre>
<p>In more serious analysis, it may be important to filter by other parameters.</p>
<h4 id="q.-can-you-find-any-other-parameters-indicating-the-quality-of-the-site">Q. Can you find any other parameters indicating the quality of the site?</h4>
<h4 id="q.-can-you-find-any-other-parameters-indicating-the-quality-of-the-call-for-a-given-individual-on-a-given-site">Q. Can you find any other parameters indicating the quality of the call for a given individual on a given site?</h4>
<h2 id="viewing-the-results-using-igv-integrative-genome-viewer">Viewing the results using IGV (Integrative Genome Viewer)</h2>
<p>In this part of the practical, we are going to use the software IGV on our local computer to visualise the alignments we created and check some of the positions where variants were called.</p>
<p>First, download the IGV software (you will need to register your email address). Unzip it to your desktop (or wherever you want it), and click it to open. If you can't click it, type, on a new terminal window:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="kw">cd</span> ~/Desktop/IGV_2_3_35/
<span class="kw">./igv.sh</span></code></pre>
<p>Doing this should open IGV.</p>
<p>We then need to get a copy the reference assembly fasta and some of the alignments (and their index files) from the cluster to your PC. Make sure you transfer BAMs from the two different colony types.</p>
<p>To run IGV, you need to define a genome file, which you have to create from the fasta alignment (Genome &gt; Genomes from file, then choose the assembly fasta file.</p>
<p>You can loads some of the BAMS and the VCF file you produced.</p>
<h4 id="q.-has-bcftoolsmpileup-recovered-the-same-positions-as-igv">Q. Has bcftools/mpileup recovered the same positions as IGV?</h4>
<h4 id="q.-do-you-think-our-filtering-was-effective">Q. Do you think our filtering was effective?</h4>
<h2 id="simple-analysis-of-the-variants">Simple analysis of the variants</h2>
<p>In this section we are going to analyse the genotypes we identified. We are going to use the software MeV to create a heat map of the genotypes and to run Principal Component Analysis (PCA). MeV does not use the standard VCF. Instead, it requires a variant matrix in a particular, non-standard format: one column for the scaffold name, one column with the variant position, and a column for each sample with the genotype.</p>
<p>Run the following script. It will read in the VCF and transform it to the genotype matrix in the required format:</p>
<pre class="sourceCode bash"><code class="sourceCode bash"><span class="co"># Create a soft link to the script</span>
<span class="kw">ln</span> -sf /data/SBCS-MSc-BioInf/2015-10-practical/map_call/vcf_to_mev.sh .
<span class="co"># Run script to create variant matrix</span>
<span class="kw">./vcf_to_mev.sh</span> filtered_variant_calls.vcf <span class="kw">&gt;</span> mev_var.txt</code></pre>
<h4 id="bonus-q-do-at-home-if-you-have-time.-one-of-the-most-common-and-tedious-jobs-for-bioinformaticians-is-to-transform-data-from-one-format-to-another.-can-you-write-your-own-script-to-create-the-variant-matrix-for-mev">Bonus Q (do at home if you have time). One of the most common (and tedious) jobs for Bioinformaticians is to transform data from one format to another. Can you write your own script to create the variant matrix for MeV?</h4>
<p>Now download the variant matrix from the cluster to your PC.</p>
<p>If you don't have MEV installed, download it <a href="http://sourceforge.net/projects/mev-tm4/files/">here</a>.</p>
<p>Open the application and import the data (under File). A heat map is produced automatically.</p>
<h4 id="q.-analyse-the-variant-calls">Q. Analyse the variant calls:</h4>
<p>In MeV, perform hierarchical clustering of the data and do a PCA. Can you find any consistent difference between B and b samples in our data?</p>
</body>
</html>
