# Practicals for Sept/Oct 2020 Genome Bioinformatics module at qmul.

## Prerequisites (we'll cover this sepeartely)

We suppose that you are at ease:

 * in the command line. For a refresher, try the SIB's [UNIX fundamentals online course](http://edu.isb-sib.ch/course/view.php?id=82).
 * with R. The [swirl() library/course](http://swirlstats.com) can help:
     * follow the *R programming* self-led course (skip the sections *Simulation* and *Dates and Times*)
     * follow at least the *ggplot* section of the *Exploratory_Data_Analysis* swirl course



## Practicals

* [Short read cleaning](./reference_genome/read-cleaning): Illumina short read cleaning
* [Reads to genome](./reference_genome/assembly): genome assembly, quality control
* [Gene prediction](./reference_genome/prediction): gene prediction, quality control
* Population sequencing to genotypes to population genetics statistics:
     * [Mapping reads, calling variants, visualizing variant calls.](./population_genetics/map_call)
     * [Analysing variants: PCA, measuring Differentiation & Diversity.](./population_genetics/popgen)


## Computers

Computers should be set up as necessary to do all of this. It may alternatively be possible to do all of this on Apocrita or on QMUL's Linux VDI (Virtual Desktop). It could also be feasible to do some of this stuff on a your favorite cluster or your local machine. However, it will be complicated. Among other things you'll need to install a bunch of [software](./software) and [data](./data). 


## Authors/Credits

The initial version of this practical was put together by
    * [Yannick Wurm](http://wurmlab.com) [@yannick__](http://twitter.com/yannick__)
    * [Oksana Riba-Grognuz](https://www.linkedin.com/in/oksana80)' contributions for the 2012 edition of this course
 
 It was heavily heavily heavily revised and improved thanks to efforts and new content by 
   * [Rodrigo Pracana](https://wurmlab.github.io/team/rpracana/)
   * Anurag Priyam, Carlos MartinezRuiz, Nazrath Nawz, [Robert Waterhouse](http://www.rmwaterhouse.org/), [Bruno Vieira](http://wurmlab.github.io/team/bmpvieira)

Special thanks to Tom King, Tom Bradford, Simon Butcher, Sam Court, ITS and ITSResearch for setting up computers and clusters at QMUL so they just work. 



## Things used in other versions of this course: 
	
* ~~[Inferring population sizes and gene flow.](./msmc/msmc-tutorial/guide) (Credit Stefan Schiffels [@stschiff](http://twitter.com/stschiff))~~
* ~~Gene expression  ( Credit Julien Roux [www](http://www.unil.ch/dee/home/menuinst/people/post-docs--associates/dr-julien-roux.html) [@_julien_roux](http://twitter.com/_julien_roux)):~~
     * ~~[From raw sequencing data to transcript expression levels.](./rnaseq/TP1)~~
     * ~~[Gene-level clustering and differential expression analysis.](./rnaseq/TP2)~~

