
<!-- saved from url=(0084)https://docs.google.com/document/pub?id=1ybFu5JaqGqu48jgjj-GVjZhYo4tb1fVX8X08W_Ro9QE -->
<html><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><script type="text/javascript">
(function(){(function(){function d(a){this.t={};this.tick=function(a,b,c){this.t[a]=[void 0!=c?c:(new Date).getTime(),b]};this.tick("start",null,a)}var a=new d;window.jstiming={Timer:d,load:a};if(window.performance&&window.performance.timing){var b=window.performance.timing,a=window.jstiming.load,c=b.navigationStart,b=b.responseStart;0<c&&b>=c&&(a.tick("_wtsrt",void 0,c),a.tick("wtsrt_","_wtsrt",b))}try{a=null,window.chrome&&window.chrome.csi&&(a=Math.floor(window.chrome.csi().pageT)),null==a&&window.gtbExternal&&
(a=window.gtbExternal.pageT()),null==a&&window.external&&(a=window.external.pageT),a&&(window.jstiming.pt=a)}catch(e){}})();})()
</script><script type="text/javascript">var KX_timer = new window.jstiming.Timer(); KX_timer.name = 'published';</script><title>Population genomics.RNA-seq.docx</title><style type="text/css">
      body {
        font-family: arial, sans, sans-serif;
        margin: 0;
      }

      iframe {
        border: 0;
        frameborder: 0;
        height: 100%;
        width: 100%;
      }

      #header, #footer {
        background: #f0f0f0;
        padding: 10px 10px;
      }

      #header {
        border-bottom: 1px #ccc solid;
      }

      #footer {
        border-top: 1px #ccc solid;
        border-bottom: 1px #ccc solid;
        font-size: 13;
      }

      #contents {
        margin: 6px;
      }

      .dash {
        padding: 0 6px;
      }
    </style></head><body><div id="header">Population genomics.RNA-seq.docx</div><div id="contents"><style type="text/css">@import url('https://themes.googleusercontent.com/fonts/css?kit=MSSLfUayeNh9PW3ng9UWrqo0P1CSBNc3gBWclSzSx0c');ol{margin:0;padding:0}.c14{max-width:487.3pt;background-color:#ffffff;padding:72pt 54pt 72pt 54pt}.c4{font-size:10pt;font-family:"Courier New"}.c9{color:inherit;text-decoration:inherit}.c12{color:#1155cc;text-decoration:underline}.c5{font-size:11pt;font-family:"Courier New"}.c0{height:12pt;direction:ltr}.c6{background-color:#ffff00;text-decoration:line-through}.c3{color:#0000ff;text-decoration:underline}.c7{font-size:10pt;font-family:"Times New Roman"}.c1{direction:ltr}.c13{font-weight:bold}.c10{font-size:10pt}.c2{font-style:italic}.c8{text-indent:36pt}.c11{margin-left:36pt}.title{padding-top:12pt;line-height:1.0;text-align:center;color:#000000;font-size:16pt;font-family:"Arial";font-weight:bold;padding-bottom:3pt}.subtitle{padding-top:0pt;line-height:1.0;text-align:center;color:#000000;font-size:12pt;font-family:"Arial";padding-bottom:3pt}li{color:#000000;font-size:12pt;font-family:"Cambria"}p{color:#000000;font-size:12pt;margin:0;font-family:"Cambria"}h1{padding-top:24pt;line-height:1.0;text-align:left;color:#345a8a;font-size:16pt;font-family:"Calibri";font-weight:bold;padding-bottom:0pt}h2{padding-top:10pt;line-height:1.0;text-align:left;color:#4f81bd;font-size:13pt;font-family:"Calibri";font-weight:bold;padding-bottom:0pt}h3{padding-top:10pt;line-height:1.0;text-align:left;color:#4f81bd;font-size:12pt;font-family:"Calibri";font-weight:bold;padding-bottom:0pt}h4{padding-top:12pt;line-height:1.0;text-align:left;color:#000000;font-size:14pt;font-family:"Cambria";font-weight:bold;padding-bottom:3pt}h5{padding-top:12pt;line-height:1.0;text-align:left;color:#000000;font-style:italic;font-size:13pt;font-family:"Cambria";font-weight:bold;padding-bottom:3pt}h6{padding-top:12pt;line-height:1.0;text-align:left;color:#000000;font-size:11pt;font-family:"Cambria";font-weight:bold;padding-bottom:3pt}</style><h1 class="c1"><span>Population genomics I: SNPs from RNA-seq</span></h1><p class="c1"><span>Eyal Privman</span></p><p class="c1"><span class="c12"><a class="c9" href="http://goo.gl/6tFdI">http://goo.gl/6tFdI</a></span></p><p class="c0"><span></span></p><p class="c1"><span>In this practical we will use Illumina transcriptome data (“RNA-seq”) to discover and genotype SNPs (Single Nucleotide Polymorphisms). We will use the same RNA-seq data that we mapped to the reference genome using TopHat. </span><span>Remember </span><span>that </span><span>these samples were chosen for a differential expression study, so this is not a “proper” population sample.</span><span>&nbsp;We will use the </span><span class="c3"><a class="c9" href="http://www.broadinstitute.org/gsa/wiki/index.php/Main_Page">GATK</a></span><span>&nbsp;(installed on the Vital-IT system) to identify sequence positions where several different reads show two alternative bases. (Also can detect insertion/deletion polymorphisms)</span></p><p class="c0"><span></span></p><p class="c1"><span>Start, as usual, by logging on to the Vital-IT system and sourcing the relevant “bashrc” file:</span></p><p class="c0"><span></span></p><p class="c1"><span class="c4">$ ssh -t </span><span class="c4 c13">username</span><span class="c4">@prd.vital-it.ch ssh dee-serv0</span><span class="c4 c13">X</span><span class="c7"><br></span><span class="c4">$ cd /scratch/cluster/weekly/</span><span class="c4 c13">username</span></p><p class="c1"><span class="c4">$ source /scratch/cluster/monthly/eprivman/summer2012/GATK.bashrc</span></p><p class="c0"><span class="c4"></span></p><p class="c1"><span>The GATK bashrc defines paths to the required Java programs.</span></p><h3 class="c1"><span>A &nbsp;Pre-process bam files</span></h3><p class="c1"><span>We will use the </span><span class="c3"><a class="c9" href="http://samtools.sourceforge.net/">SAMtools</a></span><span>&nbsp;and </span><span class="c3"><a class="c9" href="http://picard.sourceforge.net/">Picard</a></span><span>&nbsp;packages to pre-process the TopHat output files before running GATK. Use the SAMtools manual page on:</span></p><p class="c1"><span class="c12"><a class="c9" href="http://samtools.sourceforge.net/samtools.shtml">http://samtools.sourceforge.net/samtools.shtml</a></span></p><p class="c0"><span></span></p><p class="c1"><span>1</span><span>.</span><span>&nbsp;We will first process TopHat alignments of M000B RNA-seq lane stored in directory “TopHat”. </span><span>Use </span><span class="c4">AddOrReplaceReadGroups</span><span>&nbsp;from the Picard package to add required read group (RG) information </span><span>to</span><span>&nbsp;the BAM file. The Java command line is:</span></p><p class="c1"><span class="c4">$ cd TopHat</span></p><p class="c1"><span class="c4">$ myDir=M000B</span></p><p class="c1"><span class="c4">$ java </span><span class="c4">-Xmx1G</span><span class="c4">&nbsp;-jar $PICARD_HOME/AddOrReplaceReadGroups.jar I="$myDir"/accepted_hits.bam O="$myDir"/accepted_hits.rg.bam SORT_ORDER=coordinate RGID="$myDir" RGLB="$myDir" RGPL=illumina RGSM="$myDir" RGPU="$myDir"</span></p><p class="c0"><span></span></p><p class="c1"><span>2</span><span>. Use </span><span class="c5">samtools index</span><span>&nbsp;to index this BAM file (TopHat outputs a sorted </span><span>BAM)</span><span>. To see a </span><span>short help type “</span><span class="c5">samtools index</span><span>”.</span></p><p class="c0"><span></span></p><p class="c1"><span>3</span><span>. </span><span>Download the BAM file, index, genome fasta (</span><span>“GenomeSubset/Assembly/SINV_subset_1.fa”</span><span>) and cuffmerge </span><span>gtf file (“TopHat_full/merged_asm/merged.gtf”) to your computer using scp. Launch</span><span>&nbsp;IGV and import a new genome (</span><span>download genome fasta as fasta file and gtf as gene file</span><span>), then load bam file.</span><span>&nbsp;</span></p><p class="c0"><span></span></p><p class="c1 c11"><span class="c2">Q: </span><span class="c2">Take a look at position 760191 on SIgn00002</span><span class="c2">. Would you say it’s a SNP?</span></p><p class="c0"><span></span></p><p class="c1"><span>4</span><span>. Repeat the process for M350B.</span></p><p class="c0"><span></span></p><p class="c1 c11"><span class="c2">Q: </span><span class="c2">Compare positions 760187 and 760191. </span><span class="c2">Would you say that these are SNPs? Would you say so given just the M000B data?</span></p><p class="c0"><span></span></p><p class="c1"><span>5</span><span>. Use </span><span class="c5">samtools merge</span><span>&nbsp;</span><span>to merge the two sorted BAM files. </span><span>You will need to combine their headers:</span></p><p class="c1"><span class="c4">$ samtools view -H M000B/accepted_hits.bam &gt; header.sam</span></p><p class="c1"><span>Open this file in an editor and add these two lines:</span></p><p class="c1"><span class="c4">@RG&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ID:M000B&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SM:M000B</span></p><p class="c1"><span class="c4">@RG&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;ID:M350B&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;SM:M350B</span></p><p class="c1"><span>(Note: fields are separated by tab characters. If you can’t insert them as tabs you can copy a tab character from a previous line)</span></p><p class="c0"><span class="c4"></span></p><p class="c1"><span class="c4">$ samtools cat -h header.sam -o merge.accepted_hits.rg.bam M000B/accepted_hits.rg.bam M350B/accepted_hits.rg.bam</span></p><p class="c0"><span></span></p><p class="c1"><span>6. &nbsp;Sort and index this merged BAM file.</span></p><p class="c0"><span class="c4"></span></p><p class="c1"><span>7. &nbsp;(Optional) Write a script to process also one queen and one worker sample (you can base it on the runTopHat.sh script). Merge them too.</span></p><p class="c0"><span class="c4"></span></p><p class="c1"><span class="c10">(For extended description of best practices recommended by the GAT</span><span class="c10">K </span><span class="c10">people: </span><span class="c12 c10"><a class="c9" href="http://www.broadinstitute.org/gsa/wiki/index.php/Best_Practice_Variant_Detection_with_the_GATK_v2">http://www.broadinstitute.org/gsa/wiki/index.php/Best_Practice_Variant_Detection_with_the_GATK_v2</a></span><span class="c10">)</span></p><h3 class="c1"><span>B &nbsp;SNP detection and genotyping</span></h3><p class="c0"><span></span></p><p class="c1"><span>1. </span><span>Use </span><span class="c3"><a class="c9" href="http://www.broadinstitute.org/gsa/wiki/index.php/Main_Page">GATK</a></span><span>&nbsp;to discover and genotype SNPs. Use the help information:</span></p><p class="c1"><span class="c4">java </span><span class="c4">-Xmx1G </span><span class="c4">-jar $GATK_HOME/GenomeAnalysisTK.jar --help</span></p><p class="c1"><span>You will need to use the genomic sequence </span><span>“</span><span>GenomeSubset/Assembly/SINV_subset_1.f</span><span>a”</span><span>.</span><span>&nbsp;</span><span>Run </span><span class="c5">UnifiedGenotyper</span><span>&nbsp;on each sample separately and on the merged data. </span><span>Ex</span><span>a</span><span>mine the output V</span><span>CF files.</span></p><p class="c1"><span>Hint: download these files and analyze them locally.</span></p><p class="c0"><span></span></p><p class="c1 c8"><span class="c2">Q: </span><span class="c2">In which ana</span><span class="c2">lyses are the two SNP found? (hint: </span><span class="c2">grep 7601</span><span class="c2">)</span></p><p class="c0"><span></span></p><p class="c1"><span>2. Count the number of SNPs in the separate analysis of each sample.</span></p><p class="c0"><span></span></p><p class="c1 c8"><span class="c2">Q: Are there different numbers of SNPs found in different samples? Why?</span></p><p class="c0"><span></span></p><p class="c1"><span>3. Count the number of SNPs in each scaffold.</span></p><p class="c0"><span></span></p><p class="c1 c8"><span class="c2">Q: Which regions of the genome have more SNPs?</span></p><p class="c1 c8"><span class="c2">Q: You have more power for SNP detection in highly-expressed genes than in other parts of the genome. How can you control for this?</span></p><p class="c0"><span class="c4"></span></p><p class="c0"><span class="c4"></span></p><p class="c0"><span class="c6"></span></p></div><div id="footer"><span>Published by <a target="_blank" title="Learn more about Google Docs" href="https://docs.google.com/">Google Docs</a></span><span class="dash">–</span><a href="https://docs.google.com/abuse?id=1ybFu5JaqGqu48jgjj-GVjZhYo4tb1fVX8X08W_Ro9QE">Report Abuse</a><span class="dash">–</span><span>Updated automatically every 5 minutes</span></div><script type="text/javascript">
(function(){if(window.jstiming){window.jstiming.c={};window.jstiming.f=1;var e=function(b,a,d){var c=b.t[a],g=b.t.start;if(c&&(g||d))return c=b.t[a][0],void 0!=d?g=d:g=g[0],c-g},l=function(b,a,d){var c="";window.jstiming.pt&&(c+="&srt="+window.jstiming.pt,delete window.jstiming.pt);try{window.external&&window.external.tran?c+="&tran="+window.external.tran:window.gtbExternal&&window.gtbExternal.tran?c+="&tran="+window.gtbExternal.tran():window.chrome&&window.chrome.csi&&(c+="&tran="+window.chrome.csi().tran)}catch(g){}var f=
window.chrome;if(f&&(f=f.loadTimes)){f().wasFetchedViaSpdy&&(c+="&p=s");if(f().wasNpnNegotiated){var c=c+"&npn=1",i=f().npnNegotiatedProtocol;i&&(c+="&npnv="+(encodeURIComponent||escape)(i))}f().wasAlternateProtocolAvailable&&(c+="&apa=1")}var j=b.t,p=j.start,f=[],i=[],h;for(h in j)if("start"!=h&&0!=h.indexOf("_")){var k=j[h][1];k?j[k]&&i.push(h+"."+e(b,h,j[k][0])):p&&f.push(h+"."+e(b,h))}delete j.start;if(a)for(var m in a)c+="&"+m+"="+a[m];(a=d)||(a="https:"==document.location.protocol?"https://csi.gstatic.com/csi":
"http://csi.gstatic.com/csi");return[a,"?v=3","&s="+(window.jstiming.sn||"_s")+"&action=",b.name,i.length?"&it="+i.join(","):"","",c,"&rt=",f.join(",")].join("")};window.jstiming.getReportUri=l;var n=function(b,a,d){b=l(b,a,d);if(!b)return"";var a=new Image,c=window.jstiming.f++;window.jstiming.c[c]=a;a.onload=a.onerror=function(){window.jstiming&&delete window.jstiming.c[c]};a.src=b;a=null;return b};window.jstiming.report=function(b,a,d){if("prerender"==document.webkitVisibilityState){var c=!1,g=
function(){if(!c){a?a.prerender="1":a={prerender:"1"};var f;"prerender"==document.webkitVisibilityState?f=!1:(n(b,a,d),f=!0);f&&(c=!0,document.removeEventListener("webkitvisibilitychange",g,!1))}};document.addEventListener("webkitvisibilitychange",g,!1);return""}return n(b,a,d)};var o=function(b,a,d,c){return 0<d?(c?b.tick(a,c,d):b.tick(a,"",d),!0):!1};window.jstiming.getNavTiming=function(b){if(window.a&&window.a.e){var a=window.a.e;o(b,"_dns",a.j)&&o(b,"dns_",a.i,"_dns");o(b,"_con",a.h)&&o(b,"con_",
a.g,"_con");o(b,"_req",a.n)&&o(b,"req_",a.b,"_req");o(b,"_rcv",a.b)&&o(b,"rcv_",a.o,"_rcv");if(o(b,"_ns",a.d)){o(b,"ntsrt_",a.b,"_ns");o(b,"nsfs_",a.k,"_ns");var d=window.external&&window.external.startE;!d&&(window.chrome&&window.chrome.csi)&&(d=Math.floor(window.chrome.csi().startE));d&&(o(b,"_se",d),o(b,"sens_",a.d,"_se"));o(b,"ntplt0_",a.m,"_ns");o(b,"ntplt1_",a.l,"_ns")}}}};})()
</script><script type="text/javascript">KX_timer.tick('tl'); if (document.location.protocol == 'https:') {window.jstiming.report(KX_timer, undefined , 'https://gg.google.com/csi');} else {window.jstiming.report(KX_timer);}</script></body><link rel="stylesheet" type="text/css" href="data:text/css,"></html>